File: kotlin/src/main/kotlin/com/github/jntakpe/aoc2021/days/day2/Day2.kt
package com.github.jntakpe.aoc2021.days.day2

import com.github.jntakpe.aoc2021.shared.Day
import com.github.jntakpe.aoc2021.shared.readInputLines
import kotlin.math.abs

object Day2 : Day {

    override val input: List<Move> = readInputLines(2).map { it.parseMove() }

    override fun part1() = input.fold(Position()) { acc, cur -> acc.apply { move(cur) } }.result()

    override fun part2() = input.fold(Position()) { acc, cur -> acc.apply { aim(cur) } }.result()

    private fun String.parseMove() = split(' ').let { Move(Direction.valueOf(it.first().uppercase()), it.last().toInt()) }

    enum class Direction {
        FORWARD,
        DOWN,
        UP
    }

    data class Move(val direction: Direction, val units: Int)

    class Position {

        private var horizontal: Int = 0
        private var depth: Int = 0
        private var aim: Int = 0

        fun move(move: Move) {
            when (move.direction) {
                Direction.FORWARD -> horizontal += move.units
                Direction.UP -> depth += move.units
                Direction.DOWN -> depth -= move.units
            }
        }

        fun aim(move: Move) {
            when (move.direction) {
                Direction.FORWARD -> {
                    horizontal += move.units
                    depth += aim * move.units
                }
                Direction.UP -> {
                    aim -= move.units
                }
                Direction.DOWN -> {
                    aim += move.units
                }
            }
        }

        fun result() = abs(horizontal * depth)
    }
}

File: rust/src/bin/day2.rs
use aoc_2021::shared::{Day, read_input_lines};

use crate::Move::{Down, Forward, Up};

fn main() {
    Day2 {
        input: read_input_lines(2)
    }.run()
}

struct Day2 {
    input: Vec<String>,
}

enum Move {
    Forward(isize),
    Up(isize),
    Down(isize),
}

struct Position {
    horizontal: isize,
    depth: isize,
    aim: isize,
}

impl Position {
    fn new() -> Position {
        Position { horizontal: 0, depth: 0, aim: 0 }
    }

    fn shift(&mut self, shift: &Move) {
        match shift {
            Forward(u) => self.horizontal += u,
            Up(u) => self.depth += u,
            Down(u) => self.depth -= u
        }
    }

    fn aim(&mut self, aim: &Move) {
        match aim {
            Forward(u) => {
                self.horizontal += u;
                self.depth += self.aim * u
            }
            Up(u) => self.aim -= u,
            Down(u) => self.aim += u
        }
    }

    fn result(&self) -> usize {
        (&self.horizontal * &self.depth.abs()) as usize
    }
}

fn parse(raw: &str) -> Move {
    let split: Vec<&str> = raw.split(" ").collect();
    let (direction, unit) = (split[0], split[1].parse::<isize>().unwrap());
    match direction {
        "forward" => Forward(unit),
        "up" => Up(unit),
        "down" => Down(unit),
        _ => panic!("Unknown direction {}", direction)
    }
}

impl Day for Day2 {
    fn part1(&self) -> usize {
        let moves: Vec<Move> = self.input.iter().map(|s| parse(s.as_str())).collect();
        let position = moves.iter().fold(Position::new(), |mut acc, cur| {
            acc.shift(cur);
            acc
        });
        position.result()
    }

    fn part2(&self) -> usize {
        let moves: Vec<Move> = self.input.iter().map(|s| parse(s.as_str())).collect();
        let position = moves.iter().fold(Position::new(), |mut acc, cur| {
            acc.aim(cur);
            acc
        });
        position.result()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn part1_sample() {
        assert_eq!(sample_day().part1(), 150)
    }

    #[test]
    fn part2_sample() {
        assert_eq!(sample_day().part2(), 900)
    }

    fn sample_day() -> Day2 {
        Day2 {
            input: vec![
                String::from("forward 5"),
                String::from("down 5"),
                String::from("forward 8"),
                String::from("up 3"),
                String::from("down 8"),
                String::from("forward 2"),
            ]
        }
    }
}
